ocaml/typing/uniqueness_analysis.ml

```diff
@@ -437,19 +553,329 @@ end = struct
+
+  type t
+
+  type old_tag =
+    | Old_tag_unknown
+    | Old_tag_was of Tag.t
+    | Old_tag_mutated of access_order
+
+  type error =
+    | Changed_tag of
+        { old_tag : old_tag;
+          new_tag : Tag.t
+        }
+
+  exception Error of error
+
+  (** Unknown tag *)
+  val empty : t
+
+  (** Overwrite using a certain tag *)
+  val overwrite_tag : Tag.t -> t
+
+  (** Indicate a mutation (which invalidates all tags we have learned
+      and will learn). If there is also an overwrite, we fail. *)
+  val mutate_tag : t
+
+  (** Sequential composition *)
+  val seq : t -> t -> t
+
+  (** Non-deterministic choice *)
+  val choose : t -> t -> t
+
+  (** Parallel composition *)
+  val par : t -> t -> t
+
+  (** If we find out that a tag was mutated,
+      we need to promote this information to its children.
+      This is because a write to a mutable field can change
+      any tag that is reachable from the mutable field. *)
+  val promote_mutation_to_children : t -> t
+
+  (** We may not overwrite a tag we don't know.
+      At the end of the analysis, this function should
+      be called to ensure all overwrites are on known tags. *)
+  val check_no_remaining_overwritten_as : t -> unit
+
+  (** Accept the overwrites using tags that we have learned
+      from a pattern-match. *)
+  val match_with_learned_tags : Tag.Set.t -> t -> t
+
+  (** Assert that no overwrites were collected and fail else. *)
+  val assert_empty : t -> unit
+
+  val print : Format.formatter -> t -> unit
+end = struct
+  type tags =
+    { overwritten : Tag.Set.t Tag.Map.t;
+          (** The keys of the map are the tags of the overwrite.
+          When we learn new tags, we can remove keys from this map,
+          that correspond to the new tags. We then keep the learned
+          tags on the keys that did not match them to produce good
+          error messages. It is always sound to add elements to this map.
+          *)
+      was_mutated : bool
+    }
+
+  type t =
+    | Tags of tags
+    | Tag_was_mutated
+        (** We have discovered a mutable write.
+        This is dangerous: We might have accepted a overwrite based on the wrong tag.
+        If we detect this state, we reject all overwrites to this cell. *)
+
+  type old_tag =
+    | Old_tag_unknown
+    | Old_tag_was of Tag.t
+    | Old_tag_mutated of access_order
+
+  type error =
+    | Changed_tag of
+        { old_tag : old_tag;
+          new_tag : Tag.t
+        }
+
+  exception Error of error
+
+  let empty = Tags { overwritten = Tag.Map.empty; was_mutated = false }
+
+  let overwrite_tag tag =
+    Tags
+      { overwritten = Tag.Map.singleton tag Tag.Set.empty; was_mutated = false }
+
+  let mutate_tag = Tag_was_mutated
+
+  let union_with_remembered =
+    Tag.Map.union (fun _ t0 t1 -> Some (Tag.Set.union t0 t1))
+
+  let choose t0 t1 =
+    match t0, t1 with
+    | Tags t0, Tags t1 ->
+      Tags
+        { overwritten = union_with_remembered t0.overwritten t1.overwritten;
```

Comment:
The range of the map is just for error messages, right? Then I think this is OK. But if it's used for decision-making, this doesn't look right somehow, because an overwritten tag might encounter different learned tags on different branches, and that distinction gets lost here. Maybe it's all OK, but I don't really have a good semantic understanding of what this merge is doing.